Adding zeroing tensor for QuantizedCPU type in yaml

kwanghoon-meta opened the pull request:
Summary:
Gradient Checkpointing uses QUint8 type tensor quantizer which currently not supported in dispatch key.

By adding QuantizedCPU for zero_ operation, we enable gradient checkpointing compatible to custom backward function.

Since it is just filling zero element in a tensor, we can simply use same zero which uses CPU backend.

Test Plan: Signals

Differential Revision: D43144214



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

albanD commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <c10/core/ScalarType.h>
```
Why is this include here and not with all the other ones?

albanD commented on aten/src/ATen/native/quantized/cpu/LinearUnpackImpl.cpp:
```diff
@@ -65,13 +67,19 @@ std::tuple<at::Tensor, c10::optional<at::Tensor>> PackedLinearWeightsQnnp::
+            reinterpret_cast<int8_t*>(weight_origin.data_ptr<int8_t>());
+        w->unpackWeights(w_zero_points.data(), weight_ptr_int8);
+        // See for the subtraction 128
+        // https://www.internalfb.com/code/fbsource/[D40927291-V8]/fbcode/caffe2/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp?lines=326
```
This should link to github code

albanD commented on buckbuild.bzl:
```diff
@@ -1909,7 +1943,146 @@ def define_buck_targets(
+    # This library is strictly required to add a feature quantized gradient function with backend kernel
+    # Because custom gradient function calls backend kernel matrix multiplication for gradient computation for memory efficiency
+    # ,which is located in aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp
+    # Refer - D42593235
```
link PR

albanD commented on buckbuild.bzl:
```diff
@@ -1909,7 +1943,146 @@ def define_buck_targets(
+    # Because custom gradient function calls backend kernel matrix multiplication for gradient computation for memory efficiency
+    # ,which is located in aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp
+    # Refer - D42593235
+    aten_quantized_linear = [
```
Why does this list need to be explicit here? I would expect we already have such a list somewhere?

albanD commented on buckbuild.bzl:
```diff
@@ -1483,6 +1484,39 @@ def define_buck_targets(
         ],
     )
 
+    pt_xplat_cxx_library(
```
Why do you need these two new targets?

albanD commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,121 @@
+#ifdef USE_PYTORCH_QNNPACK
+TORCH_LIBRARY_IMPL(quantized, Autograd, m) {
+  m.impl(
+      TORCH_SELECTIVE_NAME("quantized::linear_dynamic"),
```
Is this function in native_functions.yaml? If so all of this can just be codegened.

kwanghoon-meta replied:
No, it is here https://www.internalfb.com/code/fbsource/[536eda60b18586b17be36889d4d09718ccf76d5b]/xplat/caffe2/aten/src/ATen/native/quantized/library.cpp?lines=153

albanD replied:
Ok ok sad....

albanD commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,121 @@
+  static auto op = at::Dispatcher::singleton()
+                       .findSchemaOrThrow("quantized::linear_dynamic", "")
+                       .typed<decltype(packed_linear_weight_grad)>();
+  auto output = op.call(input, packed_weight, reduce_range);
```
This should be a redispatch.

kwanghoon-meta replied:
I think this was last time I was suggested to use this way?, I'm unsure how to locate redispatch function for the scheme (can't find anywhere honestly)

albanD replied:
Yes in that link, it calls `op.redispatch` not `op.call`

albanD commented on torch/csrc/quantized/quantized_backward.cpp:
```diff
@@ -0,0 +1,121 @@
+using namespace generated::details;
+using namespace at;
+// Gradient into argument 1 (`input`). TODO: Support bias as well.
+struct TORCH_API PackedLinearWeightDynamicBackward : public TraceableFunction {
```
You shouldn't write custom Node by hand. Use a custom Function to tie a fw and bw together: https://pytorch.org/tutorials/advanced/cpp_autograd.html#using-custom-autograd-function-in-c

In this case, the fw should just be the redispatch call like you do below.

kimishpatel replied:
where is the fw call btw? I only see bw implementation here.

kwanghoon-meta replied:
@albanD I'm honestly unsure if it is a feasible to implement this way, we lost an access to original weight with quantized tensor has a backend QNNPACK, it frees up weight tensor.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L335

albanD replied:
> where is the fw call btw?

The fw is in the backend key. The Autograd key is only responsible for setting up the graph.

albanD replied:
> I'm honestly unsure if it is a feasible to implement this way, we lost an access to original weight tensor

I'm not sure what that changes? The signature of the custom Function is the same as packed_linear_weight_grad (you want to add the keyset to redispatch).
And from there you can save anything you need the same way you do with this one.

kwanghoon-meta committed:
Adding zeroing tensor for QuantizedCPU type in yaml (#94467)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/94467

[Gradient Checkpointing](https://github.com/cybertronai/gradient-checkpointing) uses [QUint8](https://www.internalfb.com/code/fbsource/[eaf3851611559446b54c0f84c7256220b290c020]/xplat/langtech/pyspeech/src/model/NeuralTransducerCheckpointed.cpp?lines=128-141) type tensor quantizer which currently not supported in dispatch key.

By adding QuantizedCPU for zero_ operation, we enable gradient checkpointing compatible to custom backward function.

Since it is just filling zero element in a tensor, we can simply use same zero which uses CPU backend.

Some post : https://fb.workplace.com/groups/1144215345733672/permalink/2465687553586438/

this operator with QuantizedCPU type tensor needs to support QUINT8 tensor with zero_ operation
https://www.internalfb.com/code/fbsource/[66b4ccaddf152fe7544695f7cd62256bde98ed99]/fbcode/caffe2/torch/csrc/autograd/custom_function.h?lines=345

Test Plan: Signals

Differential Revision: D43144214

fbshipit-source-id: 7b1aff489939f6e5d48684455ba16278d2b2d6e9
